### PR TITLE
Task Center init

### PR DESCRIPTION
**Pull request by AhmedSoliman**
Task Center init

This introduces a central system to manage long-running and background restate async tasks.
The core of this proposal is to help us lean more towards spawning self-contained tasks that
are addressable, trackable, and cancellable, than current deep future poll trees.

It also allows nice possibilities like:
- Limited structured concurrency (no auto-waiting for children though)
- Graceful cancellations/shutdown reduces the risk of cancellation-unsafe drops
- Potentially less memory, deep nested future state machines become shallower.
- A single place where we can auto-propagate tracing context, flag tasks by priority, schedule tasks on different runtime, provide observability of what kind of tasks are running, etc.
- Scoped tasks allow scoped cancellation. (cooperatively cancel all tasks for a partition id, or a specific tenant in the future, or be specific and filter specific kinds of tasks)
- Limit concurrency of certain tasks by kind, partition, or tenant, etc.
- Distributing tasks among multiple tokio runtimes based on the task kind
- Support for different abort-policy based on the task kind.

A the moment, I only migrated a small pieces of our code to this system to make the merging with #1180 easier.
Once #1180 is merged, I'll move the rest of our services and bifrost to use it.

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/1181).
* #1196
* #1195
* #1194
* __->__ #1181

### CODE REVIEW

**Review by tillrohrmann — COMMENTED** [review]
Thanks for creating the `TaskCenter` @AhmedSoliman. I think it looks really cool and simplifies the process of wiring tasks up with their cancellation signals. I also like that one has control over task groups (by `kind` or `partition_id`). I also like the idea of later being able to run different tasks on different Tokio runtimes depending on their priority. Hence, I think this is a useful abstraction :-)

The one thing I am not super fond of is the usage of globals. If I am not mistaken (hopefully), then this is not strictly required, though.

One thing to point out: Once one buys into the `TaskCenter` abstraction (e.g. by using the `cancel_watcher`), there is no way to run a component w/o the `TaskCenter`. Knowing whether a component requires a `TaskCenter` or not, is then on the developer. Technically it is the same as with `tokio`.

File: crates/node/src/task_center.rs
Comment (AhmedSoliman) [reply crates/node/src/task_center.rs:102]: The static is not public so that task itself won't be able to read this. Accessors like `current_task_id()` and `current_task_kind()` expose the bits we need.

File: crates/node/src/task_center.rs
Comment (AhmedSoliman) [reply crates/node/src/task_center.rs:357]: I'll remove the internal timeout and leave it up to callers.

File: crates/node/src/task_center.rs
Comment (AhmedSoliman) [reply crates/node/src/task_center.rs:293]: Doesn't `shutdown_node_on_failure` option achieve this goal?

File: crates/node/src/task_center.rs
Comment (AhmedSoliman) [reply crates/node/src/task_center.rs:29]: While I believe that it adds unnecessary friction and a small performance overhead, I'll move most of statics to an inner Arc.

File: crates/node/src/task_center.rs
Comment (AhmedSoliman) [reply crates/node/src/task_center.rs:401]: Not necessary, I will remove the timeout as per the other comment.

File: crates/node/src/task_center.rs
Comment (AhmedSoliman) [reply crates/node/src/task_center.rs:269]: Yeah, let's address those as they show up.

File: server/src/main.rs
Comment (AhmedSoliman) [inline server/src/main.rs:192]: This is not necessary but I left it here to as a visual indicator that we will drop tokio before exiting. This can be removed in the future.

File: crates/node/Cargo.toml
Comment (tillrohrmann) [inline crates/node/Cargo.toml:47]: Is this dependency still needed?

File: crates/node/Cargo.toml
Comment (tillrohrmann) [inline crates/node/Cargo.toml:67]: Should we make this a workspace dependency?

File: crates/task-center/src/lib.rs
Comment (tillrohrmann) [inline crates/task-center/src/lib.rs:70]: Is the comment still correct wrt to the timeout?

File: crates/task-center/src/lib.rs
Comment (tillrohrmann) [inline crates/task-center/src/lib.rs:89]: Should we warn if the exit code is non-zero?

```suggestion
        if exit_code == 0 {
            info!("** Shutdown requested");
        } else {
            warn!("** Shutdown requested");
        }
```

File: crates/task-center/src/lib.rs
Comment (tillrohrmann) [inline crates/task-center/src/lib.rs:97]: Does `track_caller` help with reporting stack traces in case of panics?

File: crates/task-center/src/lib.rs
Comment (tillrohrmann) [inline crates/task-center/src/lib.rs:161]: If we check after `spawn_unchecked` whether `inner.shutdown_requested` is true, then we could also stop those tasks that were started after a potential shutdown.

File: crates/task-center/src/lib.rs
Comment (tillrohrmann) [inline crates/task-center/src/lib.rs:167]: Is this a method we want to expose to users (so us)?

File: crates/task-center/src/lib.rs
Comment (tillrohrmann) [inline crates/task-center/src/lib.rs:193]: Really cool 🤩

File: crates/task-center/src/lib.rs
Comment (tillrohrmann) [inline crates/task-center/src/lib.rs:251]: Does `MetadataSync tasks` align with `TaskKind::RpcServer`?

File: crates/task-center/src/lib.rs
Comment (tillrohrmann) [inline crates/task-center/src/lib.rs:547]: Is this a left-over?

File: crates/task-center/src/lib.rs
Comment (tillrohrmann) [inline crates/task-center/src/lib.rs:500]: Do I understand this invariant correctly that the cancellation token will be canceled if the corresponding task is shut down? Also if the task shuts down normally?

**Review by tillrohrmann — APPROVED** [review]
Thanks for creating this cool piece of work @AhmedSoliman. The changes look good to me. +1 for merging once #1180 is merged (sorry for being selfish here).

File: crates/task-center/src/lib.rs
Comment (tillrohrmann) [inline crates/task-center/src/lib.rs:164]:
```suggestion
    #[track_caller]
```

**Review by tillrohrmann — APPROVED** [review]
Great work @AhmedSoliman. LGTM. +1 for merging :-)

**Pull request merged**
